src/common/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::future::Future;
+use std::num::NonZeroU64;
+use std::ops::Deref;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
+use std::sync::{Arc, LazyLock};
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+
+use arc_swap::ArcSwap;
+use parking_lot::Mutex;
+use pin_project_lite::pin_project;
+use risingwave_common::catalog::TableId;
+use risingwave_common::metrics::LabelGuardedUintGaugeVec;
+use risingwave_common::monitor::GLOBAL_METRICS_REGISTRY;
+use risingwave_common_metrics::{
+    register_guarded_uint_gauge_vec_with_registry, LabelGuardedUintGauge,
+};
+use tokio::sync::oneshot;
+use tokio::time::Sleep;
+
+static METRICS: LazyLock<LabelGuardedUintGaugeVec<1>> = LazyLock::new(|| {
+    register_guarded_uint_gauge_vec_with_registry!(
+        "backfill_rate_limit_bytes",
+        "backfill rate limit bytes per second",
+        &["table_id"],
+        &GLOBAL_METRICS_REGISTRY
+    )
+    .unwrap()
+});
+
+pin_project! {
+    #[derive(Debug)]
+    #[project = DelayProj]
+    pub enum Delay {
+        Noop,
+        Sleep{#[pin] sleep: Sleep},
+        Wait{#[pin] rx: oneshot::Receiver<()> },
+        Infinite,
+    }
+}
+
+impl Delay {
+    pub fn new(duration: Duration) -> Self {
+        match duration {
+            Duration::ZERO => Self::Noop,
+            Duration::MAX => Self::Infinite,
+            dur => Self::Sleep {
+                sleep: tokio::time::sleep(dur),
+            },
+        }
+    }
+}
+
+impl From<Duration> for Delay {
+    fn from(value: Duration) -> Self {
+        Self::new(value)
+    }
+}
+
+impl Future for Delay {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project() {
+            DelayProj::Noop => Poll::Ready(()),
+            DelayProj::Sleep { sleep } => sleep.poll(cx),
+            DelayProj::Wait { rx } => rx.poll(cx).map(|_| ()),
+            DelayProj::Infinite => Poll::Pending,
+        }
+    }
+}
+
+/// Rate limit policy.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RateLimit {
+    /// Rate limit disabled.
+    Disabled,
```

Comment:
What about `Unlimited`?

Reply:
There was a discussion for this. I'm okay with either one. https://github.com/risingwavelabs/risingwave/issues/19857#issuecomment-2550864990